src/lib/compress.c

```diff
@@ -55,530 +55,250 @@
-#include "sound/compress_offload.h"
 #include "tinycompress/tinycompress.h"
-
-#define COMPR_ERR_MAX 128
-
-/* Default maximum time we will wait in a poll() - 20 seconds */
-#define DEFAULT_MAX_POLL_WAIT_MS    20000
+#include "tinycompress/compress_ops.h"
 
 struct compress {
-	int fd;
-	unsigned int flags;
-	char error[COMPR_ERR_MAX];
-	struct compr_config *config;
-	int running;
-	int max_poll_wait_ms;
-	int nonblocking;
-	unsigned int gapless_metadata;
-	unsigned int next_track;
+	struct compress_ops *ops;
+	void *data;
+	void *dl_hdl;
 };
 
-static int oops(struct compress *compress, int e, const char *fmt, ...)
-{
-	va_list ap;
-	int sz;
-
-	va_start(ap, fmt);
-	vsnprintf(compress->error, COMPR_ERR_MAX, fmt, ap);
-	va_end(ap);
-	sz = strlen(compress->error);
-
-	snprintf(compress->error + sz, COMPR_ERR_MAX - sz,
-		": %s", strerror(e));
-	errno = e;
-
-	return -1;
-}
+extern struct compress_ops compress_hw_ops;
 
 const char *compress_get_error(struct compress *compress)
 {
-	return compress->error;
+	return compress->ops->get_error(compress->data);
 }
-static struct compress bad_compress = {
-	.fd = -1,
-};
 
 int is_compress_running(struct compress *compress)
 {
-	return ((compress->fd > 0) && compress->running) ? 1 : 0;
+	return compress->ops->is_compress_running(compress->data);
 }
 
 int is_compress_ready(struct compress *compress)
 {
-	return (compress->fd > 0) ? 1 : 0;
+	return compress->ops->is_compress_ready(compress->data);
 }
 
-static int get_compress_version(struct compress *compress)
+struct compress *compress_open(unsigned int card, unsigned int device,
+		unsigned int flags, struct compr_config *config)
 {
-	int version = 0;
+	struct compress *compress;
+	char name[128];
 
-	if (ioctl(compress->fd, SNDRV_COMPRESS_IOCTL_VERSION, &version)) {
-		oops(compress, errno, "cant read version");
-		return -1;
+	snprintf(name, 128, "hw:%u,%u", card, device);
+	compress = calloc(1, sizeof(struct compress));
+	if (!compress)
+		return NULL;
+
+	compress->ops = &compress_hw_ops;
+	compress->data =  compress->ops->open_by_name(name, flags, config);
+	if (compress->data == NULL) {
+		free(compress);
+		return NULL;
 	}
-	return version;
+	return compress;
 }
 
-static bool _is_codec_type_supported(int fd, struct snd_codec *codec)
+static int populate_compress_plugin_ops(struct compress *compress, const char *name)
 {
-	struct snd_compr_caps caps;
-	bool found = false;
-	unsigned int i;
-
-	if (ioctl(fd, SNDRV_COMPRESS_GET_CAPS, &caps)) {
-		oops(&bad_compress, errno, "cannot get device caps");
-		return false;
+	unsigned int ret = -1;
+	char *token, *token_saveptr;
+	char *compr_name;
+	char lib_name[128] = "lib";
+	void *dl_hdl;
+	const char *err = NULL;
+
+	token = strdup(name);
+	compr_name = strtok_r(token, ":", &token_saveptr);
+	free(token);
+
+	strncat(lib_name, compr_name, strlen(lib_name) + strlen(compr_name) + 1);
+	strncat(lib_name, ".so", strlen(lib_name) + strlen(".so") + 1);
+
+	dl_hdl = dlopen(lib_name, RTLD_NOW);
```

Comment:
We may discuss, if the plugins should be in another directory (not the system wide). Also, the plugin libraries may contain the special prefix or so..

For example, in alsa-lib, we have the separate directory for plugins like '/usr/lib64/alsa-lib' and the plugins has names like 'libasound_module_pcm_pulse.so'.

Reply:
Probably we can have some macro TINYCOMPRESS_PLUGIN_LIB path default set /usr/lib/tinycompress-lib/ which can be updated from Makefile and plugin library name like libtinycompress_module_<plugin_name>.so . Is that ok ?

Reply:
Yep, it sounds good. @vinodkoul - do you agree?

Reply:
@perexg yes good idea.
The tinycompress in installed in /usr/lib/ should we add plugin default patch as that or create a tinycompress directory for these plugins?
What does Alsa-lib do, we should have a common approach in both..

Reply:
and on the plugin name, yes lgtm

Reply:
In alsa-lib, default plugin path is /usr/lib/alsa-lib/ . We can have default plugin path as /usr/lib/tinycompress-lib/.

Reply:
Done. Please review.

Reply:
> In alsa-lib, default plugin path is /usr/lib/alsa-lib/ . We can have default plugin path as /usr/lib/tinycompress-lib/.

Ok sounds good to me.. @perexg ?